### ISSUE TITLE
Encoded URL Handling / Redirect Bug

### ISSUE THREAD

**Issue by FStelzer** [oauth2-proxy/oauth2-proxy#844]
oauth2-proxy seems to sanitize / unescape parts of the proxied url and send a redirect when the backend expects the url exactly as specified. This breaks some backends that use url encoded parameters.
Our current example is a rabbitmq-server admin frontend. RabbitMQ by default has a virtual host called "/" and generates some urls in the form of:
/api/vhosts/%2F/permissions

oauth2-proxy will then send me a HTTP 301 to /api/vhosts/permissions for this url which the backend then not recognizes.

## Expected Behavior

No redirect / rewriting of urls from oauth2-proxy

## Current Behavior

Calling /api/vhosts/%2F/permissions results in a HTTP 301 to /api/vhosts/permissions from oauth2-proxy

## Possible Solution

Oauth2-proxy should not touch / rewrite / redirect the specified url when passing to the backend proxy in any way or at least have such behaviour configurable.

## Steps to Reproduce (for bugs)

Issue a request to an url like:
/FOO/%2F/BAR/
and no matter what's behine the proxy you will receive a HTTP 301 to
/FOO/BAR/

- Version used: 6.1.1

**Comment by JoelSpeed** [comment]
Is this redirect behaviour present only after authentication flows or any time you make this request? I don't see what would be returning a 301 🤔 

Have you checked your upstream servers logs to ensure there's nothing there that's causing it to send a 301, could be some mismatch between host headers or something like that I'm thinking

**Comment by FStelzer** [comment]
I'm only seeing this after authentication.
i did a tcpdump on the backend again to verify. My browser is requesting:
/api/vhosts/%2F/permissions
and receives a HTTP 301
Location: /api/vhosts/permissions

The backend never even see's this request. The first thing hitting the backend is the redirected /api/vhosts/permissions (which produces a 404 because of the missing %2F in between)

**Comment by JoelSpeed** [comment]
Ahh, I think I've heard of this before. 

I believe there is a problem with the Go HTTP server implementation that doesn't like escaped characters in paths like this and tries to get rid of them.

Do you see a log line for this request? I would assume the OAuth2 Proxy would be logging that it is returning a 301?

**Comment by FStelzer** [comment]
correct. i see the 301 and then the 404:

Nov 05 14:52:35 HOST oauth2-proxy[5530]: XXX:45906 - XXUSER [2020/11/05 14:52:35] HOST:8005 GET - "/api/vhosts/%2F/permissions" HTTP/1.1 "Mozilla/5.0 (X11; Fedora; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.121 Safari/537.36" 301 58 0.000
Nov 05 14:52:35 HOST oauth2-proxy[5530]: XXX:45906 - XXUSER [2020/11/05 14:52:35] HOST:8005 GET / "/api/vhosts/permissions" HTTP/1.1 "Mozilla/5.0 (X11; Fedora; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.121 Safari/537.36" 404 49 0.002

And yes this seems to be a known go thing, i found a few similar bugs in other projects.

**Comment by FStelzer** [comment]
Yes this is still very relevant! If someone familiar with the oauth2-proxy codebase (@JoelSpeed ?) could give me some directions on where the request flows through the code or where to place a possible fix i can probably patch it myself. But a bit of code reading in the proxy did not give me any leads.

**Comment by JoelSpeed** [comment]
I'm not entirely sure where this is coming from still, but looking at some previous conversations it seems it may have something to do with the serve mux. If that's the case, it is likely to be somewhere around here https://github.com/oauth2-proxy/oauth2-proxy/blob/master/pkg/upstream/proxy.go

Perhaps we can add some test cases into this package and see if we get the same results as a starting point?

**Comment by FStelzer** [comment]
This incredibly simple change actually fixes it... 
I just use the escaped raw path for all upstream proxy requests. The Pull request ist just a PoC but i think if i could restrict the raw path usage to just the upstream and not also static and file handlers the i think this would be fine.
I think the proxy should touch/modify the request as little as possible.

**Comment by FStelzer** [comment]
still work in progress

**Comment by eXeDK** [comment]
Any news on this @FStelzer ? We are also running into this when trying to but rabbitmq behind oauth2-proxy

**Comment by FStelzer** [comment]
just updated my pull request with the new config changes. if these are ok i can add at least the global setting of passing the raw path easily. setting it per upstream requires a larger change of changing the mux and also upstream mux changes.

**Comment by eXeDK** [comment]
@FStelzer do you see any synergy with https://github.com/oauth2-proxy/oauth2-proxy/pull/1128 ?

**Comment by FStelzer** [comment]
still relevant

**Comment by FStelzer** [comment]
A new pull request is waiting for review

**Comment by FStelzer** [comment]
it was fixed. you'll need to use the alpha config and set
upstreamConfig:
  proxyRawPath: true
